### DISCUSSION TITLE
DYNAMO CODING

### DISCUSSION THREAD

**Discussion by alezet2001** [community/community#122005] [Programming Help]
### Body

Hey! I have a problem with my Dynamo-code. I want to read an Excel-file to update parameters in revit. I want every Type of a family (for example doors) get the same parameter Values. And I want to reuse it for projects so I can´t go on Element-ID for example. Any ideas?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by setojan** [comment]
I would like to understand how exactly an Excel table is structured for this task. The goal is to input specific parameter values into the types of certain families and then transfer these families to other projects.

What about processing family files in a directory? And then importing these families into projects while overwriting the parameters?

**Comment by tmilost** [comment]
Hi @alezet2001  

You can absolutely achieve this in Dynamo—here’s how to do it in a way that’s reusable for any Revit project, without relying on Element IDs.

## Step-by-Step Solution

### 1. Structure Your Excel File  
Prepare your Excel in this format, where each row lists a family name, type name, and parameter values:

| Family Name | Type Name    | ParameterA | ParameterB |  
|-------------|--------------|------------|------------|  
| Door        | 910x2100mm   | Value A    | Value B    |  
| Door        | 1000x2100mm  | Value C    | Value D    |  

Keep the column headers for clarity. This structure ensures your mapping works across projects, as it uses family and type names (not element ids).

### 2. Dynamo Workflow

**A. Read Excel Data**  
- Use `Excel.ReadFromFile` to import your sheet.  
- Pass the output through `List.Transpose` if needed (to flip rows/columns), and remove headers with `List.DropItems`.

**B. Collect All Family Types**  
- Use `Categories` and select your category (e.g., Doors).
- Use `All Elements of Category` to collect all family instances.
- Use `Element.Type` to get the family types.
- Use `Element.Name` for type names, and `Element.GetParameterValueByName` for family names (pass “Family Name” as input).

**C. Match Family and Type Names from Excel to Revit Types**  
- For each type in your model, match its family name and type name with those from Excel.
- You can use combinations of `==`, `&&`, `List.Map`, or a small Python script to match the pairs.

**D. Set Parameter Values**  
- For matched types, use `SetParameterByName` nodes to assign each parameter from Excel to its corresponding family type in Revit.

---

## Example Python Script Node for Matching (Optional)

If you want to do the matching with a Python script node in Dynamo:

```python
# IN[0]=ExcelData, IN[1]=Dynamo FamilyTypes, IN[2]=TypeNames, IN[3]=FamilyNames
matched_types = []
matched_values = []

for revit_ft, t_name, f_name in zip(IN[1], IN[2], IN[3]):
    for row in IN[0]:
        if row[0] == f_name and row[1] == t_name:
            matched_types.append(revit_ft)
            matched_values.append(row[2])  # Adjust for your parameter column
            break
OUT = matched_types, matched_values
```
Feed the output into `SetParameterByName`.

---

## Tips for Reusability

- **No Element Ids** – Use family and type names for mapping, not Revit IDs.
- **Consistent Naming** – Make sure type/family names stay consistent across projects.
- **Flexible Excel** – You can reuse the same Excel logic for doors, windows, furniture, etc., by changing the category in Dynamo.

---

### References:
- [Dynamo Forums: Type Parameter Updating](https://forum.dynamobim.com/)

---

**Comment by andriiafanasiev** [comment]
You can map Excel rows to family type names instead of Element-IDs, then use Dynamo to loop over all types in a family and assign the corresponding parameter values. To reuse across projects, make sure the mapping relies on type names or another consistent identifier, and consider processing the family files in a folder first, updating parameters, then loading them into the project.

If this helps, please mark this comment as the Accepted Answer.
